Mage/src/main/java/mage/game/combat/CombatGroup.java

```diff
@@ -544,9 +549,14 @@ private void multiAttackerDamage(boolean first, Game game) {
 
     private void defenderDamage(Permanent attacker, int amount, Game game) {
         if (this.defenderIsPlaneswalker) {
-            Permanent defender = game.getPermanent(defenderId);
-            if (defender != null) {
-                defender.markDamage(amount, attacker.getId(), null, game, true, true);
+            Permanent planeswalker = game.getPermanent(defenderId);
+            if (planeswalker != null) {
+                //apply damage to player from TrampleOverPlaneswalkersAbility (for ThrastaTempestsRoar in set MH2)
+                if (hasTrampleOverPlaneswalkers(attacker) && (amount > planeswalker.getCounters(game).getCount(CounterType.LOYALTY))) {
+                    Player defendingPlayer = game.getPlayer(defendingPlayerId);
+                    defendingPlayer.damage((amount - planeswalker.getCounters(game).getCount(CounterType.LOYALTY)), attacker.getId(), null, game, true, true);
```

Comment:
I changed this part to make it look like the normal damage trample code in newest commit. Also, I tried to use defenderDamage instead of defendingPlayer.damage() but it did not work correctly. I tried this:
```
                if (hasTrampleOverPlaneswalkers(attacker)) {
                    int lethalDamage = planeswalker.getLethalDamage(attacker.getId(), game);
                    if (lethalDamage >= amount) {
                        planeswalker.markDamage(amount, attacker.getId(), null, game, true, true);
                    } else {
                        planeswalker.markDamage(amount, attacker.getId(), null, game, true, true);
                        amount -= lethalDamage;
                        if (amount > 0) {
                            defenderDamage(amount, attacker, game);
                        }
                    }
```
but it did not work, and damage was applied to planeswalker but not player. I thought it was because the defenderId needs to be set to defendingPlayerId  before defenderDamage is called again. I tried that too, and it made player take correct damage, but planeswalker was not damaged after combat. 

I just pushed new commit that uses getLethalDamage to find damage, but then uses player.damage() to apply the trample damage. Check the new commit to see

Reply:
Try to add additional param to `defenderDamage` calls like `boolean damageToDefenderController`

``` java
    /**
     * Do damage to attacked player or planeswalker
     *
     * @param game
     * @param attacker
     * @param amount
     * @param damageToDefenderController excess damage to defender's controller (example: trample over planeswalker)
     */
    private void defenderDamage(Game game, Permanent attacker, int amount, boolean damageToDefenderController) {
        UUID affectedDefenderId = this.defenderId;
        if (damageToDefenderController) {
            affectedDefenderId = game.getControllerId(this.defenderId);
        }

        // on planeswalker
        Permanent planeswalker = game.getPermanent(defenderId);
        if (planeswalker != null) {
            // apply excess damage from "trample over planeswaslkers" ability (example: Thrasta, Tempest's Roar)
            if (hasTrampleOverPlaneswalkers(attacker)) {
                int lethalDamage = planeswalker.getLethalDamage(attacker.getId(), game);
                if (lethalDamage >= amount) {
                    // normal damage
                    planeswalker.markDamage(amount, attacker.getId(), null, game, true, true);
                } else {
                    // damage with excess (additional damage to planeswalker's controller)
                    planeswalker.markDamage(lethalDamage, attacker.getId(), null, game, true, true);
                    amount -= lethalDamage;
                    if (amount > 0) {
                        defenderDamage(game, attacker, amount, true);
                    }
                }
            } else {
                // normal damage
                planeswalker.markDamage(amount, attacker.getId(), null, game, true, true);
            }
        }

        // on player
        Player defender = game.getPlayer(affectedDefenderId);
        if (defender != null) {
            defender.damage(amount, attacker.getId(), null, game, true, true);
        }
    }
```

Reply:
I added that in latest commit and refactored in extra parameter. Seems to work well.